src/operator/split.rs

```diff
@@ -0,0 +1,138 @@
+    let transform = |mut token: String| {
+        let mut trimmed = true;
+        while trimmed {
+            trimmed = false;
+            for (start, end) in closures {
+                if token.starts_with(start) && token.ends_with(end) {
+                    token = token[start.len()..token.len() - end.len()].to_string();
+                    trimmed = true;
+                }
+            }
+        }
+
+        map(token)
+    };
+
+    while index < input.len() {
+        let mut found_separator = false;
+        let remaining = &input[index..];
+        if remaining.starts_with(separator) {
+            found_separator = true;
+        } else {
+            let mut found = false;
+            for (start, end) in closures {
+                if !remaining.starts_with(start) {
+                    continue;
+                }
+
+                if let Some(end_index) = remaining[start.len()..].find(end) {
+                    index += start.len() + end_index + end.len();
+                    found = true;
+                    break;
+                }
+            }
+
+            if !found {
+                index += 1;
+            }
+        }
+
+        if found_separator {
+            if index != prev_index {
+                tokens.push(transform(input[prev_index..index].to_string()));
+            }
+            index += separator.len();
+            prev_index = index;
+        }
+    }
+
+    if index != prev_index {
+        tokens.push(transform(input[prev_index..index].to_string()));
+    }
+
+    tokens
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn split_with_separator_and_closures_string(
+        input: &str,
+        separator: &str,
+        closures: &HashMap<&'static str, &'static str>,
+    ) -> Vec<String> {
+        split_with_separator_and_closures(input, separator, closures, |s| s)
+    }
+
+    #[test]
+    fn split_works() {
+        assert_eq!(
+            split_with_separator_and_closures_string("power hello", " ", &DEFAULT_CLOSURES),
+            vec!["power", "hello"],
+        );
+        assert_eq!(
+            split_with_separator_and_closures_string("morecomplicated", "ecomp", &DEFAULT_CLOSURES),
+            vec!["mor", "licated"],
+        );
+        assert_eq!(
+            split_with_separator_and_closures_string("owmmowmow", "ow", &DEFAULT_CLOSURES),
+            vec!["mm", "m"],
+        );
+    }
+
+    #[test]
+    fn split_with_closures_works() {
+        assert_eq!(
+            split_with_separator_and_closures_string(
+                "power hello \"good bye\"",
+                " ",
+                &DEFAULT_CLOSURES
+            ),
+            vec!["power", "hello", "good bye"],
+        );
+        assert_eq!(
+            split_with_separator_and_closures_string(
+                "more'ecomp'licated",
+                "ecomp",
+                &DEFAULT_CLOSURES
+            ),
+            vec!["more'ecomp'licated"],
```

Comment:
a more sensible result in this case would be `more', 'licated`

Reply:
If the closures don't enclose a separator then the case for:
```
hello 'good bye'
```

doesn't group 'good bye' when being split by spaces:
```
hello, good, bye
```

Maybe the difference with this example is it's not being split by whitespace?  The set of closures to check for could be different depending on the input (no closures if non-whitespace).

Or maybe the difference is that the quoted group is purely a separator. An example like:
```
hello' 'good
```

does seems like it should be split into:
```
hello', 'good
```

Since it feels like the user maybe made a typing mistake (`hello' 'good bye'`). I'm not sure what other tools do in this case. 

I feel like one possible solution is to avoid consuming the closed group if it starts with or ends with the separator. But IMO keeping the logic / rule-set simple is preferred.